File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    max = 0;

    diameterOfBinaryTree(root) {
        if (root === null) return 0;
        this.dfs(root);
        return this.max;
    }

    dfs(node) {
        if(node === null) return 0;
        var left = this.dfs(node.left);
        var right = this.dfs(node.right);
        var currentPath = left+right;
        this.max = Math.max(currentPath,this.max);
        return Math.max(left,right) + 1;
    }
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        Queue<TreeNode> q = new LinkedList();
        List<Integer> result = new ArrayList<>();

        if (root == null) {
            return List.of();
        }

        q.offer(root);

        while (q.size() > 0) {
            int levelSize = q.size();
            for (int i = 0; i < levelSize; i++) {
                TreeNode curr = q.poll();

                if (curr.left != null) {
                    q.offer(curr.left);
                }

                if (curr.right != null) {
                    q.offer(curr.right);
                }
                if (i == levelSize - 1) {
                    result.add(curr.val);
                }
            }
        }
        return result;
    }
}
